### DISCUSSION TITLE
Deployed wasm Edge function fails at runtime with "TypeError: file system access error (code 42)"

### DISCUSSION THREAD

**Discussion by audkar** [supabase/supabase#13737] [Questions] [answered]
I followed the steps in https://deno.com/blog/wasmbuild to build Rust code to a Deno compatible wasm binary. 

Basically code:
```
import { instantiate } from "./lib/rs_lib.generated.js";

const { add } = await instantiate();

console.log(add(1, 2));
```

The edge function fails at runtime after deploying to Supabase cloud
```
supabase functions deploy wasm
```
Making a request return `501` status code

Logs contains:
```
TypeError: file system access error (code 42), id:cbafe93a-99ac-4655-844d-cf5d7e317210
    at async mainFetch (ext:deno_fetch/26_fetch.js:266:12)
    at async fetch (ext:deno_fetch/26_fetch.js:490:7)
    at async instantiateModule (file:///src/main.ts:385:38)
    at async file:///src/main.ts:334:35
    at async instantiate (file:///src/main.ts:323:13)
    at async file:///src/main.ts:400:24
``` 

---

When running locally everything works.

`supabase functions serve  wasm` gives output:
```
<...>
Check file:///home/deno/functions/wasm/index.ts
3
Listening on http://localhost:8000/
<...>
```

**Comment by gregnr** [comment]
Thanks for reporting @audkar. @laktek any thoughts on this one?

**Comment by samurai00** [comment]
It seems to be an issue related to Deno file read permission policies.
If I run test using deno run command:

```shell
┌ ⚠️  Deno requests read access to "/Users/xxx/Workspace/supabase/supabase/functions/_shared/test_lib_bg.wasm".
├ Requested by `Deno.readFile()` API.
├ Run again with --allow-read to bypass this prompt.
└ Allow? [y/n/A] (y = yes, allow; n = no, deny; A = allow all read permissions) >
```

**Reply by audkar** [reply]
(If this is the underlying problem) then is it possible to configure Supabase Edge Functions Runtime to somehow allow access to a specific file?

**Comment by laktek** [comment, marked as answer]
@audkar The hosted environment doesn't have access to file system. You can host your wasm file (in S3 or Supabase Storage) and use `instantiateStreaming` within hosted environment. Here's an example on how to do it https://github.com/supabase/edge-runtime/blob/main/examples/wasm-stream/index.ts

**Reply by audkar** [reply]
Thank you @laktek . It works.

But it looks strange to have an edge function that loads its runtime code from external sources. Isn't that inefficient from a runtime speed and cost perspective? I suspect each call will also count Supabase Storage egress? Are there any plans to have fully deployable wasm functions at the edge?

**Reply by laktek** [reply]
Yes, it's not an ideal developer experience right now. We have some improvements planned for better wasm module support.
